File: apps/ios/TaakHelden/App/AppState.swift
import Foundation
import Observation

@Observable
final class AppState {
    enum Route {
        case welcome
        case parentOnboarding
        case childPairing
        case childUnlock
        case childHome
    }

    var route: Route = .welcome
    var selectedChildTab: ChildTab = .mijnDag
    let environment: AppEnvironment

    var authStore: AuthStore { environment.authStore }
    var apiClient: TaakHeldenAPIClient { environment.apiClient }
    var mutationQueue: MutationQueue { environment.mutationQueue }
    var syncEngine: SyncEngine { environment.syncEngine }
    var parentGate: ParentGateCoordinator { environment.parentGate }
    var localAuth: LocalAuthenticationClient { environment.localAuth }
    var celebrationService: CelebrationService { environment.celebrationService }
    var photoBonusService: PhotoBonusService { environment.photoBonusService }
    var pushService: PushRegistrationService { environment.pushService }

    var parentMode: ParentModeStore
    var pendingParentDeepLinkSurface: ParentSurface?

    init(usePreviewData: Bool = false) {
        environment = AppEnvironment(usePreviewData: usePreviewData)
        parentMode = ParentModeStore(
            apiClient: environment.parentAPIClient,
            familyRoomClient: environment.familyRoomClient
        )
    }

    @MainActor
    func restoreSessionIfAvailable() async {
        switch authStore.restoredRoute {
        case .childHome:
            route = .childHome
        case .childUnlock:
            route = .childUnlock
        case .parentOnboarding:
            route = .parentOnboarding
        case .welcome:
            route = .welcome
        }
    }

    func openParentOnboarding() {
        route = .parentOnboarding
    }

    func openChildPairing() {
        route = .childPairing
    }

    func finishChildPairing() {
        route = .childHome
    }

    func unlockChildHome() {
        authStore.unlockChildSession()
        route = .childHome
    }

    func openParentGate(from entryPoint: ParentGateEntryPoint, preferSurface: ParentSurface? = nil) {
        if let preferSurface {
            pendingParentDeepLinkSurface = preferSurface
        }
        parentGate.openGate(from: entryPoint)
    }

    @MainActor
    func attemptLocalAuthUnlock() async -> Bool {
        do {
            let success = try await localAuth.evaluateDeviceOwner(
                reason: NSLocalizedString("parent.gate.la.reason", comment: "")
            )
            if success {
                parentGate.unlock(using: .localAuthentication)
                applyPendingDeepLinkSurface()
            }
            return success
        } catch {
            return false
        }
    }

    func unlockParentMode(using method: ParentGateUnlockMethod) {
        parentGate.unlock(using: method)
        applyPendingDeepLinkSurface()
    }

    @MainActor
    func closeParentMode() {
        parentMode.endSession()
        parentGate.closeParentMode()
        pendingParentDeepLinkSurface = nil
    }

    /// Logs out the current user and navigates to the welcome screen.
    ///
    /// Deregisters the device push token for the **departing** user before
    /// clearing sessions.  On a shared iPad, only the acting user's
    /// (apnsToken, user_id) row is removed; other profiles remain intact.
    @MainActor
    func returnToWelcome() async {
        if let token = APNSTokenStore.shared.apnsToken,
           authStore.childSession != nil || authStore.parentSession != nil {
            try? await apiClient.deregisterDevice(apnsToken: token)
        }
        authStore.clearAllSessions()
        route = .welcome
    }

    @MainActor
    func handleBackgroundPushRefresh() async {
        await syncEngine.syncNow()
        if parentGate.isParentModePresented {
            await parentMode.handleBackgroundPushRefresh()
        }
    }

    @MainActor
    func enforceParentIdleTimeoutIfNeeded() {
        guard parentGate.isParentModePresented else { return }
        if parentGate.parentSessionRequiresReauth() {
            closeParentMode()
        }
    }

    private func applyPendingDeepLinkSurface() {
        if let pendingParentDeepLinkSurface {
            parentMode.activeSurface = pendingParentDeepLinkSurface
            self.pendingParentDeepLinkSurface = nil
        }
    }
}

File: apps/android/app/src/main/kotlin/nl/taakhelden/family/ui/AppState.kt
package nl.taakhelden.family.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import nl.taakhelden.core.auth.AppRestoreRoute
import nl.taakhelden.core.gate.ParentGateEntryPoint
import nl.taakhelden.core.gate.ParentGateUnlockMethod
import nl.taakhelden.core.parent.ParentModeStore
import nl.taakhelden.core.parent.ParentSurface
import nl.taakhelden.family.AppEnvironment

enum class AppRoute {
    WELCOME,
    PARENT_ONBOARDING,
    CHILD_PAIRING,
    CHILD_UNLOCK,
    CHILD_HOME,
}

enum class ChildTab {
    MIJN_DAG,
    WINKEL,
    MIJN_STER,
}

/**
 * Top-level navigation and cross-cutting app coordination.
 *
 * A ViewModel rather than composable state so a rotation or a language change never drops
 * the parent-mode session or the realtime connection mid-approval.
 */
class AppState(val environment: AppEnvironment) : ViewModel() {

    private val _route = MutableStateFlow(AppRoute.WELCOME)
    val route: StateFlow<AppRoute> = _route.asStateFlow()

    private val _selectedChildTab = MutableStateFlow(ChildTab.MIJN_DAG)
    val selectedChildTab: StateFlow<ChildTab> = _selectedChildTab.asStateFlow()

    private var pendingParentDeepLinkSurface: ParentSurface? = null

    val parentMode: ParentModeStore = ParentModeStore(
        api = environment.parentApi,
        familyRoomClient = environment.familyRoomClient,
        scope = viewModelScope,
        openTaskCountSink = environment.preferences,
    )

    val parentGate = environment.parentGate
    val authStore = environment.authStore

    init {
        restoreSessionIfAvailable()
    }

    fun selectChildTab(tab: ChildTab) {
        _selectedChildTab.value = tab
    }

    private fun restoreSessionIfAvailable() {
        _route.value = when (authStore.restoredRoute) {
            AppRestoreRoute.CHILD_HOME -> AppRoute.CHILD_HOME
            AppRestoreRoute.CHILD_UNLOCK -> AppRoute.CHILD_UNLOCK
            AppRestoreRoute.PARENT_ONBOARDING -> AppRoute.PARENT_ONBOARDING
            AppRestoreRoute.WELCOME -> AppRoute.WELCOME
        }
    }

    fun openParentOnboarding() {
        _route.value = AppRoute.PARENT_ONBOARDING
    }

    fun openChildPairing() {
        _route.value = AppRoute.CHILD_PAIRING
    }

    fun finishChildPairing() {
        _route.value = AppRoute.CHILD_HOME
    }

    fun unlockChildHome() {
        authStore.unlockChildSession()
        _route.value = AppRoute.CHILD_HOME
    }

    fun openParentGate(
        entryPoint: ParentGateEntryPoint,
        preferSurface: ParentSurface? = null,
    ) {
        if (preferSurface != null) pendingParentDeepLinkSurface = preferSurface
        parentGate.openGate(entryPoint)
    }

    fun unlockParentMode(method: ParentGateUnlockMethod) {
        parentGate.unlock(method)
        applyPendingDeepLinkSurface()
    }

    fun closeParentMode() {
        parentMode.endSession()
        parentGate.closeParentMode()
        pendingParentDeepLinkSurface = null
    }

    /**
     * Signs the current user out.
     *
     * The push token is detached *before* the session is cleared — afterwards there is no
     * token to authenticate the de-registration with, and the departing user would keep
     * receiving this family's notifications.
     */
    fun returnToWelcome() {
        viewModelScope.launch {
            if (authStore.childSession != null || authStore.parentSession != null) {
                environment.pushService.deregisterCurrentUser()
            }
            authStore.clearAllSessions()
            _route.value = AppRoute.WELCOME
        }
    }

    /** Called when the app is backgrounded: the child home always re-locks. */
    fun handleAppBackgrounded() {
        authStore.lockChildSession()
        if (_route.value == AppRoute.CHILD_HOME) {
            _route.value = AppRoute.CHILD_UNLOCK
        }
    }

    fun handleAppForegrounded() {
        if (parentGate.isParentModePresented.value &&
            parentGate.parentSessionRequiresReauth()
        ) {
            closeParentMode()
        }
        if (_route.value == AppRoute.CHILD_HOME) {
            viewModelScope.launch { environment.syncEngine.syncNow() }
        }
        viewModelScope.launch { environment.pushService.registerIfNeeded() }
    }

    fun handleBackgroundPushRefresh() {
        viewModelScope.launch {
            environment.syncEngine.syncNow()
            if (parentGate.isParentModePresented.value) {
                parentMode.handleBackgroundPushRefresh()
            }
        }
    }

    fun handlePushDeepLink() {
        if (_route.value == AppRoute.CHILD_HOME) {
            openParentGate(
                ParentGateEntryPoint.HERO_WORDMARK_LONG_PRESS,
                preferSurface = ParentSurface.GOEDKEUREN,
            )
        }
    }

    private fun applyPendingDeepLinkSurface() {
        pendingParentDeepLinkSurface?.let { surface ->
            parentMode.setActiveSurface(surface)
            pendingParentDeepLinkSurface = null
        }
    }

    override fun onCleared() {
        super.onCleared()
        environment.familyRoomClient.disconnect()
    }

    class Factory(private val environment: AppEnvironment) : ViewModelProvider.Factory {
        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel> create(modelClass: Class<T>): T = AppState(environment) as T
    }
}
